ambry-router/src/main/java/com.github.ambry.router/CloudRouterFactory.java

```diff
@@ -0,0 +1,163 @@
+import com.github.ambry.commons.SSLFactory;
+import com.github.ambry.commons.ServerMetrics;
+import com.github.ambry.config.CloudConfig;
+import com.github.ambry.config.ClusterMapConfig;
+import com.github.ambry.config.NetworkConfig;
+import com.github.ambry.config.RouterConfig;
+import com.github.ambry.config.ServerConfig;
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.network.LocalNetworkClientFactory;
+import com.github.ambry.network.LocalRequestResponseChannel;
+import com.github.ambry.network.NetworkClientFactory;
+import com.github.ambry.network.NetworkMetrics;
+import com.github.ambry.notification.NotificationSystem;
+import com.github.ambry.protocol.AmbryRequests;
+import com.github.ambry.protocol.RequestHandlerPool;
+import com.github.ambry.server.StoreManager;
+import com.github.ambry.store.StoreKeyConverterFactory;
+import com.github.ambry.store.StoreKeyFactory;
+import com.github.ambry.utils.SystemTime;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.io.IOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link RouterFactory} that creates routers that work with cloud storage.
+ */
+public class CloudRouterFactory implements RouterFactory {
+  private static final Logger logger = LoggerFactory.getLogger(CloudRouterFactory.class);
+  private final RouterConfig routerConfig;
+  private final NonBlockingRouterMetrics routerMetrics;
+  private final ClusterMap clusterMap;
+  private final NotificationSystem notificationSystem;
+  private final AccountService accountService;
+  private final Time time;
+  private final NetworkClientFactory networkClientFactory;
+  private final KeyManagementService kms;
+  private final CryptoService cryptoService;
+  private final CryptoJobHandler cryptoJobHandler;
+  private final String defaultPartitionClass;
+  private final RequestHandlerPool requestHandlerPool;
+
+  /**
+   * Creates an instance of NonBlockingRouterFactory with the given {@code verifiableProperties},
+   * {@code clusterMap} and {@code notificationSystem}.
+   * @param verifiableProperties the in-memory properties to use to construct configurations.
+   * @param clusterMap the {@link ClusterMap} to use to determine where operations should go.
+   * @param notificationSystem the {@link NotificationSystem} to use to log operations.
+   * @param sslFactory the {@link SSLFactory} to support SSL transmissions. Required if SSL is enabled for any
+   *                   datacenters.
+   * @param accountService the {@link AccountService} to use.
+   * @throws Exception if any of the arguments are null or if instantiation of KMS or CryptoService fails
+   */
+  public CloudRouterFactory(VerifiableProperties verifiableProperties, ClusterMap clusterMap,
+      NotificationSystem notificationSystem, SSLFactory sslFactory, AccountService accountService) throws Exception {
```

Comment:
Is throwing a generic exception ok? Should we more specific?

Reply:
there are many to handle.